src/main/kotlin/frc/robot/subsystems/drive/ProfiledPosePID.kt

```diff
@@ -0,0 +1,136 @@
+package frc.robot.subsystems.drive
+
+import edu.wpi.first.math.controller.ProfiledPIDController
+import edu.wpi.first.math.geometry.Pose2d
+import edu.wpi.first.math.geometry.Rotation2d
+import edu.wpi.first.math.kinematics.ChassisSpeeds
+import edu.wpi.first.math.trajectory.TrapezoidProfile.Constraints
+import edu.wpi.first.units.Units
+import edu.wpi.first.wpilibj2.command.button.Trigger
+import frc.robot.lib.LoggedNetworkGains
+import org.littletonrobotics.junction.Logger
+import org.team5987.annotation.LoggedOutput
+
+private val gainsX = LoggedNetworkGains("gainX", 4.0)
+private val gainsY =
+    LoggedNetworkGains(
+        "gainY",
+    )
+private val gainsTheta = LoggedNetworkGains("gainTheta", 6.0)
+
+private val LINEAR_CONSTRAINTS =
+    Constraints(
+        TunerConstants.kSpeedAt12Volts.`in`(Units.MetersPerSecond),
+        Units.MetersPerSecondPerSecond.of(2.8)
+            .`in`(Units.MetersPerSecondPerSecond)
+    )
+
+private val ROTATIONAL_CONSTRAINTS =
+    Constraints(
+        TunerConstants.kMaxOmegaVelocity.`in`(Units.RadiansPerSecond),
+        Units.DegreesPerSecondPerSecond.of(360.0)
+            .`in`(Units.RadiansPerSecondPerSecond)
+    )
+
+val xController =
+    ProfiledPIDController(
+        gainsX.kP.get(),
+        gainsX.kI.get(),
+        gainsX.kD.get(),
+        LINEAR_CONSTRAINTS
+    )
+val yController =
+    ProfiledPIDController(
+        gainsY.kP.get(),
+        gainsY.kI.get(),
+        gainsY.kD.get(),
+        LINEAR_CONSTRAINTS
+    )
+
+val thetaController =
+    ProfiledPIDController(
+            gainsTheta.kP.get(),
+            gainsTheta.kI.get(),
+            gainsTheta.kD.get(),
+            ROTATIONAL_CONSTRAINTS
+        )
+        .apply { enableContinuousInput(-Math.PI, Math.PI) }
+
+fun updateProfiledPID() {
+    xController.setPID(gainsX.kP.get(), gainsX.kI.get(), gainsX.kD.get())
+    yController.setPID(gainsY.kP.get(), gainsY.kI.get(), gainsY.kD.get())
+    thetaController.setPID(
+        gainsTheta.kP.get(),
+        gainsTheta.kI.get(),
+        gainsTheta.kD.get()
+    )
+}
```

Comment:
Don't repeat yourself:
```suggestion
fun updateProfiledPID() {
    mapOf(
            xController to gainsX,
            yController to gainsY,
            thetaController to gainsTheta
        )
        .forEach { (controller, gains) ->
            controller.setPID(gains.kP.get(), gains.kI.get(), gains.kD.get())
        }
}
```

Reply:
af6651b150e2564df19926b7d7e50c926ba6b431